core/src/main/java/com/radixdlt/messaging/core/MessageDispatcher.java

```diff
@@ -137,26 +143,82 @@ CompletableFuture<Result<Unit, Cause>> send(final OutboundMessageEvent outboundM
         .findOrCreateChannel(outboundMessage.receiver())
         .thenApply(channel -> send(channel, bytes))
         .thenApply(this::updateStatistics)
-        .exceptionally(t -> completionException(t, receiver, message));
+        .thenApply(
+            result -> {
+              result.onError(err -> logSendError(message, receiver, err.toString()));
+              return result;
+            })
+        .exceptionally(
+            t -> {
+              final var cause = t.getCause() != null ? t.getCause().getMessage() : t.getMessage();
+              logSendError(message, receiver, cause);
+              return IO_ERROR.result();
+            });
+  }
+
+  private void logSendError(Message message, NodeId receiver, String cause) {
+    final RateLimiter rateLimiter;
+    final AtomicInteger countSinceLastLog;
+    if (sendErrorLogRateLimitersByReceiver.contains(receiver)) {
```

Comment:
hmm, I see a mix of thread-safe Atomics and non-thread-safe `LRUCache`, which makes me wonder - can this code be called from multiple threads? should we synchronize the cache access, or demote the AtomicInt to int?

Reply:
huh, I was pretty sure this only runs on a single thread, the P2P thread (AtomicInteger was merely a mutable container), but as you mentioned it I double checked and it's actually multi (two) threaded. Nice catch :) I'll fix that.

Reply:
done